Mathlib/LinearAlgebra/RootSystem/Hom.lean

```diff
@@ -82,31 +102,450 @@ def Hom.comp {ι₁ M₁ N₁ ι₂ M₂ N₂ : Type*} [AddCommGroup M₁] [Modu
+    indexEquiv (x * y) = indexEquiv x ∘ indexEquiv y :=
+  rfl
+
+/-- The endomorphism monoid of a root pairing. -/
+abbrev _root_.RootPairing.End (P : RootPairing ι R M N) := Hom P P
+
+/-- The weight space representation of endomorphisms -/
+def weightHom (P : RootPairing ι R M N) : End P →* (M →ₗ[R] M) where
+  toFun g := Hom.weightMap (P := P) (Q := P) g
+  map_mul' g h := by ext; simp
+  map_one' := by ext; simp
+
+lemma weightHom_injective (P : RootPairing ι R M N) : Injective (weightHom P) := by
+  intro f g hfg
+  ext x
+  · exact congrFun (congrArg DFunLike.coe hfg) x
+  · refine (LinearEquiv.injective P.toDualRight) ?_
+    simp_rw [← weight_coweight_transpose_apply]
+    exact congrFun (congrArg DFunLike.coe (congrArg LinearMap.dualMap hfg)) (P.toDualRight x)
+  · refine (Embedding.injective P.root) ?_
+    simp_rw [← root_weightMap_apply]
+    exact congrFun (congrArg DFunLike.coe hfg) (P.root x)
+
+/-- The coweight space representation of endomorphisms -/
+def coweightHom (P : RootPairing ι R M N) : End P →* (N →ₗ[R] N)ᵐᵒᵖ where
+  toFun g := MulOpposite.op (Hom.coweightMap (P := P) (Q := P) g)
+  map_mul' g h := by
+    simp only [← MulOpposite.op_mul, coweightMap_mul, LinearMap.mul_eq_comp]
+  map_one' := by
+    simp only [MulOpposite.op_eq_one_iff, coweightMap_one, LinearMap.one_eq_id]
+
+lemma coweightHom_injective (P : RootPairing ι R M N) : Injective (coweightHom P) := by
+  intro f g hfg
+  ext x
+  · dsimp [coweightHom] at hfg
+    rw [MulOpposite.op_inj] at hfg
+    have h := congrArg (LinearMap.comp (M₃ := Module.Dual R M)
+        (σ₂₃ := RingHom.id R) (P.toDualRight)) hfg
+    rw [← f.weight_coweight_transpose, ← g.weight_coweight_transpose] at h
+    have : f.weightMap = g.weightMap := by
+      haveI refl : Module.IsReflexive R M := PerfectPairing.reflexive_left P.toPerfectPairing
+      exact (Module.dualMap_dualMap_eq_iff R M (Bijective.injective
+        (Module.bijective_dual_eval R M))).mp (congrArg LinearMap.dualMap
+        ((LinearEquiv.eq_comp_toLinearMap_iff f.weightMap.dualMap g.weightMap.dualMap).mp h))
+    exact congrFun (congrArg DFunLike.coe this) x
+  · dsimp [coweightHom] at hfg
+    simp_all only [MulOpposite.op_inj]
+  · dsimp [coweightHom] at hfg
+    rw [MulOpposite.op_inj] at hfg
+    set y := f.indexEquiv x with hy
+    have : f.coweightMap (P.coroot y) = g.coweightMap (P.coroot y) := by
+      exact congrFun (congrArg DFunLike.coe hfg) (P.coroot y)
+    rw [coroot_coweightMap_apply, coroot_coweightMap_apply, Embedding.apply_eq_iff_eq, hy] at this
+    rw [Equiv.symm_apply_apply] at this
+    rw [this, Equiv.apply_symm_apply]
+
+end Hom
+
+variable {ι₂ M₂ N₂ : Type*}
+    [AddCommGroup M₂] [Module R M₂] [AddCommGroup N₂] [Module R N₂]
+    (P : RootPairing ι R M N) (Q : RootPairing ι₂ R M₂ N₂)
+
+/-- An equivalence of root pairings is a morphism where the maps of weight and coweight spaces are
+bijective.  `Equiv.toEndUnit` gives an isomorphism to invertible morphisms. -/
```

Comment:
Fully qualifying a name will make it work as a link in the docs. (Working tweak just personal preference, feel free to ignore.)
```suggestion
bijective.

See also `RootPairing.Equiv.toEndUnit`. -/
```